### PR TITLE
Remove gogoproto nullable and embed from messages.proto

### PR DESCRIPTION
**Pull request by KrishnaIyer**
#### Summary

Brace yourselves. This is a big one . Refs https://github.com/TheThingsNetwork/lorawan-stack/issues/2798
This is mostly updating usage of pointers but there are a couple of tricky areas.

#### Changes

- Remove gogoproto nullable and embed from messages.proto
- Update code and tests.

#### Testing

Existing tests should cover this.

##### Regressions

Don't expect any but you never know.

##### Notes for Reviewers

1. `ApplicationDownlinkFailed` is not used in any RPCs.
2. Please check [`json_test.go`](https://github.com/TheThingsNetwork/lorawan-stack/blob/362a70995b96c334037a5e26a214eb3d78f8445c/pkg/applicationserver/io/formatters/json_test.go#L61) @adriansmares. Since we can how have nil elements, I've removed those fields from the expected json, rather than initialise them in the message.
3. Does our [API validation work](https://github.com/TheThingsNetwork/lorawan-stack/blob/362a70995b96c334037a5e26a214eb3d78f8445c/pkg/gatewayserver/io/mqtt/mqtt.go#L331-L333) on MQTT messages that the GS MQTT frontend receives? 

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: cmd/ttn-lw-cli/commands/simulate.go
Comment (adriansmares) [inline cmd/ttn-lw-cli/commands/simulate.go]: Probably should be `.UTC()` in retrospective, and we'll use the `ttnpb` functions after https://github.com/TheThingsNetwork/lorawan-stack/pull/4889 is merged.

File: pkg/applicationserver/applicationserver.go
Comment (adriansmares) [inline pkg/applicationserver/applicationserver.go:308]: I'd prefer if we propagate the pointer (i.e. change the signature of `getLink` to take `*ttnpb.EndDeviceIdentifiers`)

I'm ok with doing this systematically in another PR if you feel this is out of scope here.

File: pkg/applicationserver/applicationserver.go
Comment (adriansmares) [inline pkg/applicationserver/applicationserver.go:704]: Same for these handlers.

File: pkg/applicationserver/applicationserver.go
Comment (adriansmares) [inline pkg/applicationserver/applicationserver.go:954]: This is panic worthy to be honest.

File: pkg/applicationserver/io/grpc/grpc.go
Comment (adriansmares) [inline pkg/applicationserver/io/grpc/grpc.go:216]: Make `skipPayloadCrypto` take pointer.

File: pkg/applicationserver/io/packages/loradms/v1/package.go
Comment (adriansmares) [inline pkg/applicationserver/io/packages/loradms/v1/package.go:74]: Pointer identifiers.

File: pkg/applicationserver/io/packages/loradms/v1/package.go
Comment (adriansmares) [inline pkg/applicationserver/io/packages/loradms/v1/package.go:150]: Pointer identifiers.

File: pkg/applicationserver/io/packages/loragls/v3/package.go
Comment (adriansmares) [inline pkg/applicationserver/io/packages/loragls/v3/package.go:70]: Pointer identifiers.

File: pkg/applicationserver/io/packages/loragls/v3/package.go
Comment (adriansmares) [inline pkg/applicationserver/io/packages/loragls/v3/package.go:81]: Pointer identifiers.

File: pkg/applicationserver/io/web/webhooks.go
Comment (adriansmares) [inline pkg/applicationserver/io/web/webhooks.go:230]: Pointer identifiers.

File: pkg/applicationserver/io/web/webhooks.go
Comment (adriansmares) [inline pkg/applicationserver/io/web/webhooks.go:251]: Pointer identifiers.

File: pkg/gatewayserver/io/mqtt/mqtt.go
Comment (adriansmares) [inline pkg/gatewayserver/io/mqtt/mqtt.go]: This is panic worthy to be honest.

File: pkg/ttnpb/application_interfaces.go
Comment (adriansmares) [inline pkg/ttnpb/application_interfaces.go]: What about the `ExtractRequestFields` ?

File: pkg/applicationserver/applicationserver.go
Comment (KrishnaIyer) [reply pkg/applicationserver/applicationserver.go:308]: Yup I do prefer it in a separate PR; https://github.com/TheThingsNetwork/lorawan-stack/issues/4896
I want to get this one merged before I get more conflicts.

File: pkg/gatewayserver/io/mqtt/mqtt.go
Comment (KrishnaIyer) [reply pkg/gatewayserver/io/mqtt/mqtt.go]: This was one of my questions in the PR
> Does our API validation work on MQTT messages that the GS MQTT frontend receives?

Where does this happen? The MQTT client publishes a marshalled byte and we unmarshal it based on the topic.

Comment (KrishnaIyer) [reply pkg/gatewayserver/io/mqtt/mqtt.go]: Is there a middleware somewhere?

Comment (adriansmares) [reply pkg/gatewayserver/io/mqtt/mqtt.go]: Let's do it for all frontends:

```diff
diff --git a/pkg/gatewayserver/io/io.go b/pkg/gatewayserver/io/io.go
index e4a53ca6d..d40dc82ba 100644
--- a/pkg/gatewayserver/io/io.go
+++ b/pkg/gatewayserver/io/io.go
@@ -202,11 +202,21 @@ var errBufferFull = errors.DefineInternal("buffer_full", "buffer is full")
 const consecutiveRepeatUpEventsInterval = time.Minute
 
 // HandleUp updates the uplink stats and sends the message to the upstream channel.
-func (c *Connection) HandleUp(up *ttnpb.UplinkMessage) error {
+func (c *Connection) HandleUp(up *ttnpb.UplinkMessage) (err error) {
 	if c.discardRepeatedUplink(up) {
 		return nil
 	}
 
+	defer func() {
+		if err != nil {
+			registerDropMessage(c.ctx, c.gateway, "uplink", err)
+		}
+	}()
+
+	if err := up.ValidateFields(); err != nil {
+		return err
+	}
+
 	var ct scheduling.ConcentratorTime
 	if up.Settings.Time != nil {
 		ct = c.scheduler.SyncWithGatewayAbsolute(up.Settings.Timestamp, up.ReceivedAt, *up.Settings.Time)
@@ -262,15 +272,23 @@ func (c *Connection) HandleUp(up *ttnpb.UplinkMessage) error {
 		atomic.StoreInt64(&c.lastUplinkTime, up.ReceivedAt.UnixNano())
 		c.notifyStatsChanged()
 	default:
-		err := errBufferFull.New()
-		registerDropMessage(c.ctx, c.gateway, "uplink", err)
-		return err
+		return errBufferFull.New()
 	}
 	return nil
 }
 
 // HandleStatus updates the status stats and sends the status to the status channel.
-func (c *Connection) HandleStatus(status *ttnpb.GatewayStatus) error {
+func (c *Connection) HandleStatus(status *ttnpb.GatewayStatus) (err error) {
+	defer func() {
+		if err != nil {
+			registerDropMessage(c.ctx, c.gateway, "status", err)
+		}
+	}()
+
+	if err := status.ValidateFields(); err != nil {
+		return err
+	}
+
 	select {
 	case <-c.ctx.Done():
 		return c.ctx.Err()
@@ -286,24 +304,30 @@ func (c *Connection) HandleStatus(status *ttnpb.GatewayStatus) error {
 			}
 		}
 	default:
-		err := errBufferFull.New()
-		registerDropMessage(c.ctx, c.gateway, "status", err)
-		return err
+		return errBufferFull.New()
 	}
 	return nil
 }
 
 // HandleTxAck sends the acknowledgment to the status channel.
-func (c *Connection) HandleTxAck(ack *ttnpb.TxAcknowledgment) error {
+func (c *Connection) HandleTxAck(ack *ttnpb.TxAcknowledgment) (err error) {
+	defer func() {
+		if err != nil {
+			registerDropMessage(c.ctx, c.gateway, "txack", err)
+		}
+	}()
+
+	if err := ack.ValidateFields(); err != nil {
+		return err
+	}
+
 	select {
 	case <-c.ctx.Done():
 		return c.ctx.Err()
 	case c.txAckCh <- ack:
 		c.notifyStatsChanged()
 	default:
-		err := errBufferFull.New()
-		registerDropMessage(c.ctx, c.gateway, "txack", err)
-		return err
+		return errBufferFull.New()
 	}
 	return nil
 }

```

Comment (KrishnaIyer) [reply pkg/gatewayserver/io/mqtt/mqtt.go]: This opened a can of worms that needed to be tamed. I think this is good now.

File: api/gateway.proto
Comment (KrishnaIyer) [inline api/gateway.proto:336]: We never validated this but the values from the TTKG and TTIG are more than 32 chars already. Extending this to pass validation.

File: pkg/gatewayserver/io/ws/lbslns/upstream.go
Comment (adriansmares) [inline pkg/gatewayserver/io/ws/lbslns/upstream.go]: Why do we need a copy ? Is not like you can `Get` twice.

File: api/messages.proto
Comment (adriansmares) [inline api/messages.proto]: I would leave this commentary out, as this is an implementation detail. The panics in the code are description enough in my opinion.

**Review by adriansmares — APPROVED** [review]

File: pkg/gatewayserver/io/ws/lbslns/upstream.go
Comment (KrishnaIyer) [reply pkg/gatewayserver/io/ws/lbslns/upstream.go]: Because now `Settings` is a pointer, it gets modified [here](https://github.com/TheThingsNetwork/lorawan-stack/blob/b2d03c3ad5a6d068c6c6596c4e1697b93ddb3993/pkg/gatewayserver/io/ws/ws_test.go#L1037).

I could change the test case but we might somewhere down the line mutate this inadvertently in the code. Better safe than sorry.

File: api/messages.proto
Comment (htdvisser) [inline api/messages.proto]: This is a breaking change. Proto names are part of the API.

File: pkg/ttnpb/application_interfaces.go
Comment (htdvisser) [inline pkg/ttnpb/application_interfaces.go]: I'm pretty sure that in `pkg/ttnpb` we alsways use `m` as receiver. Also for the other ones below.

File: api/messages.proto
Comment (KrishnaIyer) [reply api/messages.proto]: Oh I see. Ok fixed.

File: pkg/gatewayserver/io/ws/lbslns/upstream.go
Comment (adriansmares) [reply pkg/gatewayserver/io/ws/lbslns/upstream.go]: Personally I'd rather fix the test case, and in another PR/issue make the downlink tokens also erase the downlink on `Get`. 

If we ever mutate the settings somewhere down the line we need to fix that, not do a copy here.

Comment (KrishnaIyer) [reply pkg/gatewayserver/io/ws/lbslns/upstream.go]: >  Is not like you can Get twice.

Actually you can. The reason for not erasing the token on `Get` is that this is not a push-pop mechanism. For some reason I don't exactly remember, we may expect multiple ACKs with the same `Diid` and so we used to retain the Downlink. I haven't seen this happen so we can indeed pop the Downlink on a valid Tx Ack. 
I don't agree however where this copying should be. I feel like doing it at source is a good idea if we know beforehand that there are elements that may be mutated.
Idiomatically, since the `Get` doesn't pop, we can also move the copying there.

File: pkg/gatewayserver/io/mqtt/mqtt.go
Comment (adriansmares) [reply pkg/gatewayserver/io/mqtt/mqtt.go]: This check can now be dropped since the validation frontend will do it.

File: pkg/gatewayserver/io/ws/lbslns/upstream.go
Comment (adriansmares) [reply pkg/gatewayserver/io/ws/lbslns/upstream.go]: Copying in `Get` sounds good to me.

**Review by adriansmares — APPROVED** [review]

**Comment by KrishnaIyer** [comment]
This is waiting on Code owner approval. @htdvisser could you check please?

**Review by htdvisser — APPROVED** [review]
Code-owned LGTM

**Pull request merged**
